bug(node): FSL validation error when nulls at end of vector column batch

If the last value in a vector column is passed in as null, then we can get an error when it cross the FFI bridge. The root cause is an upstream bug in apache-arrow: https://github.com/apache/arrow/issues/45862

```js
let arrow = require("apache-arrow");

let array = arrow.vectorFromArray(
    [[1, 2, 3], null],
    new arrow.FixedSizeList(3, new  arrow.Field("item", new  arrow.Float32())),
)

let table = new arrow.Table({name: array});


let lancedb = require("@lancedb/lancedb");

let db = await lancedb.connect("memory://")

let db_table = await db.createTable("my_table", table);
```

```
[Error: lance error: LanceError(Arrow): Arrow error: Invalid argument error: Values length 3 is less than the length (3) multiplied by the value size (3) for FixedSizeList(Field { name: "item", data_type: Float32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, 3), /Users/runner/.cargo/git/checkouts/lance-b31243ab5673a03e/c73d717/rust/lance-datafusion/src/utils.rs:51:31  Caused by: LanceError(Arrow): Arrow error: Invalid argument error: Values length 3 is less than the length (3) multiplied by the value size (3) for FixedSizeList(Field { name: "item", data_type: Float32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, 3), /Users/runner/.cargo/git/checkouts/lance-b31243ab5673a03e/c73d717/rust/lance-datafusion/src/utils.rs:51:31] {
  code: 'GenericFailure'
}
```

Note: there is a bug in the error message, fixed here: https://github.com/apache/arrow-rs/pull/7313

--- a/nodejs/__test__/table.test.ts
+++ b/nodejs/__test__/table.test.ts
@@ -279,6 +279,15 @@ describe.each([arrow15, arrow16, arrow17, arrow18])(
       expect(res.getChild("y")?.toJSON()).toEqual([2, null, null, null]);
       expect(res.getChild("z")?.toJSON()).toEqual([null, null, 3n, 5n]);
     });
+
+    it("should handle null vectors at end of data", async () => {
+      // https://github.com/lancedb/lancedb/issues/2240
+      const data = [{ vector: [1, 2, 3] }, { vector: null }];
+      const db = await connect("memory://");
+
+      const table = await db.createTable("my_table", data);
+      expect(await table.countRows()).toEqual(2);
+    });
   },
 );
 

--- a/nodejs/lancedb/arrow.ts
+++ b/nodejs/lancedb/arrow.ts
@@ -37,10 +37,10 @@ import {
   Utf8,
   Vector,
   makeVector as arrowMakeVector,
+  vectorFromArray as badVectorFromArray,
   makeBuilder,
   makeData,
   makeTable,
-  vectorFromArray,
 } from "apache-arrow";
 import { Buffers } from "apache-arrow/data";
 import { type EmbeddingFunction } from "./embedding/embedding_function";
@@ -183,6 +183,21 @@ export class VectorColumnOptions {
 
   constructor(values?: Partial<VectorColumnOptions>) {
     Object.assign(this, values);
+  }
+}
+
+// biome-ignore lint/suspicious/noExplicitAny: skip
+function vectorFromArray(data: any, type?: DataType) {
+  // Workaround for: https://github.com/apache/arrow/issues/45862
+  // If FSL type with float
+  if (DataType.isFixedSizeList(type) && DataType.isFloat(type.valueType)) {
+    const extendedData = [...data, new Array(type.listSize).fill(0.0)];
+    const array = badVectorFromArray(extendedData, type);
+    return array.slice(0, data.length);
+  } else if (type === undefined) {
+    return badVectorFromArray(data);
+  } else {
+    return badVectorFromArray(data, type);
   }
 }
 

